rst_files/schelling.rst

```diff
@@ -143,76 +128,62 @@ Here's one solution that does the job we want. If you feel like a
 further exercise you can probably speed up some of the computations and
 then increase the number of agents.
 
-.. code-block:: julia
-  :class: test
+Setup
+-----
 
-  using Test
+.. literalinclude:: /_static/includes/deps.jl
 
 .. code-block:: julia
+    :class: test
 
-    using Plots, Random, LinearAlgebra
-    gr(fmt=:png)
+    using Test
 
-    Random.seed!(42)  # set seed for random numbers. Reproducible output
+.. code-block:: julia
 
+    using Parameters, Plots, Random
+    gr(fmt = :png)
 
 .. code-block:: julia
 
-    mutable struct Agent{TI<:Integer, TF<:AbstractFloat}
-        kind::TI
-        location::Vector{TF}
-    end
+    Random.seed!(42);  # set seed for random numbers. Reproducible output
 
-    # constructor
-    Agent(k) = Agent(k, rand(2))
+.. code-block:: julia
 
+    Agent = @with_kw (kind, location = rand(2))
 
-    function draw_location!(a)
-        a.location = rand(2)
-        nothing
-    end
+    draw_location!(a) = a.location .= rand(2)
 
     # distance is just 2 norm: uses our subtraction function
     get_distance(a, o) = norm(a.location - o.location)
 
-    function is_happy(a, others)
-        "True if sufficient number of nearest neighbors are of the same type."
+    function is_happy(a, others, neighborhood_size, preference)
```

Comment:
I think this is best written as it was before; i.e., `is_happy(agent, others)`

Reply:
I think this should be a style guide, but using enclosure functions is kind of risky and should be avoided if possible. Having all the arguments to fully characterize the function also helps for transferring parts of code.

Reply:
I think pretty much the opposite.  We have been making heavy use of closures, and should continue to do so.  The key is that you can copy the whole code inside of a function to transfer it.

If we are fixing the `neighborhood_size` and `preference` and just applying it to `a` and `others` then it is a lot more direct to use a closure.

It is also easier to read, since the reader can see what is really treated as a function argument.